xdsl/transforms/lower_riscv_func.py

```diff
@@ -0,0 +1,58 @@
+from typing import cast
+from xdsl.dialects.builtin import ModuleOp
+
+from xdsl.ir import MLContext, Operation
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriteWalker,
+    op_type_rewrite_pattern,
+    RewritePattern,
+    PatternRewriter,
+)
+from xdsl.dialects import riscv, riscv_func
+from xdsl.transforms.dead_code_elimination import dce
+
+
+class LowerSyscallOp(RewritePattern):
+    """
+    Lower SSA version of syscall, storing the optional result to a0.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: riscv_func.SyscallOp, rewriter: PatternRewriter):
+        ops: list[Operation] = []
+
+        for i, arg in enumerate(op.args):
+            ops.append(
+                riscv.MVOp(
+                    arg,
+                    rd=riscv.Register(f"a{i}"),
+                )
+            )
+
+        ops.append(riscv.LiOp(immediate=op.syscall_num, rd=riscv.Registers.A7))
```

Comment:
Not all kernels expect the syscall code to be in `a7`, freeBSD puts it in `t0` for example.

There is a whole discussion going on, and currently there is no "one standard" for syscall lowering. We should probably document the syscall ABI we implement somewhere. (I did not know about all of this when I wrote the emulator, and the emulator is relatively close to what "standard" Linux does, afaik).

Reply:
Do you think this is something that should be done as part of this PR? Maybe I could your comment to the code, so that when there's unexpected behaviour we can try and rectify it? Right now I've gone with your emulator's "ABI" or at least my understanding of it...

Reply:
I added a comment